### PR TITLE
Quantized Interpolate Kernel(upsample_nearest2d)

### PR DESCRIPTION
**Pull request by lly-zero-one**
In this PR, we implemented the support of quantized interpolate with upsample_nearest2d case. 

import torch, time

for dtype in [torch.qint8, torch.quint8, torch.qint32]:
    print('****', str(dtype), '*****')
    x = torch.rand(1, 56, 56, 256)

    q_x = torch.quantize_per_tensor(x, 0.5, 1, dtype)
    q_x = q_x.permute([0, 3, 1, 2])

    x = x.permute([0, 3, 1, 2])

    NITER = 100

    s = time.time()
    for i in range(NITER):
        # float_out = torch.nn.functional.avg_pool2d(x, kernel_size=5, stride=None, padding=0)
        # float_out = torch.nn.functional.adaptive_avg_pool2d(x, output_size=5)
        float_out = torch.nn.functional.interpolate(x, size=5, scale_factor=None, mode="nearest", align_corners=None)
    time_per_iter_float = (time.time() - s) / NITER

    s = time.time()
    for i in range(NITER):
        # quant_out = torch.nn.quantized.functional.avg_pool2d(q_x, kernel_size=5, stride=None, padding=0)
        # quant_out = torch.nn.quantized.functional.adaptive_avg_pool2d(q_x, output_size=5)
        quant_out = torch.nn.quantized.functional.interpolate(q_x, size=5, scale_factor=None, mode="nearest", align_corners=None)
    time_per_iter_quant = (time.time() - s) / NITER

    ref_quantized = torch.quantize_per_tensor(float_out, 0.5, 1, dtype)
    #  torch.testing.assert_allclose(ref_quantized.dequantize(), quant_out.dequantize())

    print('time/iter ms (float)', 'time/iter ms (quant)', 'quant/float', sep='\t')
    print(time_per_iter_float * 1000, time_per_iter_quant * 1000, time_per_iter_quant / time_per_iter_float, sep='\t')

    bytes_float = (x.numel() + float_out.numel()) * x.element_size()
    bytes_quant = (q_x.numel() + quant_out.numel()) * q_x.element_size()

    float_bw_gbps = bytes_float / time_per_iter_float / 1e9
    quant_bw_gbps = bytes_quant / time_per_iter_quant / 1e9

    print('GB/s float', 'GB/s quant', sep='\t')
    print(float_bw_gbps, quant_bw_gbps, sep='\t')

=========without special handling of NHWC layout=============
**** torch.qint8 *****
time/iter ms (float)    time/iter ms (quant)    quant/float
2.08712100982666        2.1624231338500977      1.0360794240817361
GB/s float      GB/s quant
1.5508750976872339      0.37421723220248165
**** torch.quint8 *****
time/iter ms (float)    time/iter ms (quant)    quant/float
2.056601047515869       2.184889316558838       1.0623787823107091
GB/s float      GB/s quant
1.573890086222483       0.3703693335250963
**** torch.qint32 *****
time/iter ms (float)    time/iter ms (quant)    quant/float
2.0152783393859863      2.067704200744629       1.0260142037623525
GB/s float      GB/s quant
1.6061622539873104      1.5654386148823074

=========with special handling of NHWC layout=============
**** torch.qint8 *****
time/iter ms (float)    time/iter ms (quant)    quant/float
2.044649124145508       0.009250640869140625    0.004524317038018256
GB/s float      GB/s quant
1.5830902044636819      87.47675014597938
**** torch.quint8 *****
time/iter ms (float)    time/iter ms (quant)    quant/float
2.049403190612793       0.009107589721679688    0.004444020465761265
GB/s float      GB/s quant
1.579417859221808       88.8507305147644
**** torch.qint32 *****
time/iter ms (float)    time/iter ms (quant)    quant/float
2.0601415634155273      0.01062631607055664     0.0051580513976618066
GB/s float      GB/s quant
1.5711852318699757      304.6082930818039

### CODE REVIEW

**Comment by llyfacebook** [comment]
TODO: for nhwc layout support

File: aten/src/ATen/native/quantized/cpu/qupsample_nearest2d.cpp [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+          nearest_neighbor_compute_source_index(height_scale, h2, input_height);
+
+      for (int64_t w2 = 0; w2 < output_width; ++w2) {
+        const int64_t w1 =
+            nearest_neighbor_compute_source_index(width_scale, w2, input_width);
+
+        const auto* pos1 = &i_p[(h1 * input_width + w1)*channels];
+        auto* pos2 = &o_p[(h2 * output_width + w2)*channels];
+
+        for (int64_t c = 0; c < channels; ++c) {
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/qupsample_nearest2d.cpp]: This is definitely going to be auto-vectorizable. See https://gcc.godbolt.org/z/QTr8AE vs https://gcc.godbolt.org/z/S1ag2z. Maybe we should move this to `kernels/`

File: aten/src/ATen/native/quantized/cpu/qupsample_nearest2d.cpp [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+#include <ATen/native/SortingUtils.h>
+#include <ATen/native/TensorIterator.h>
+#include <ATen/native/UpSample.h>
+#include <ATen/native/cpu/Loops.h>
+#include <ATen/native/quantized/cpu/quantized_ops.h>
+#include <ATen/quantized/Quantizer.h>
+
+namespace at {
+namespace native {
+namespace {} // namespace
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/qupsample_nearest2d.cpp]: What's this for?

**Review by jamesr66a — COMMENTED** [review]
Generally looks good on the whole, but please add some performance benchmarks comparing quantized to float and to machine peak

File: aten/src/ATen/native/quantized/cpu/qupsample_nearest2d.cpp [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+        pos2[0] = pos1[0];
+        pos1 += input_height * input_width;
+        pos2 += output_height * output_width;
+      }
+    }
+  }
+}
+
+template <typename scalar_t>
+static void upsample_nearest2d_out_frame_nhwc(
```
[the remark this answers is not in the document]
Comment (llyfacebook) [reply aten/src/ATen/native/quantized/cpu/qupsample_nearest2d.cpp]: I did see significant performance improvement for nhwc layout input. I think one might be the contiguous call cost. Another one might be memory access.

File: aten/src/ATen/native/quantized/cpu/qupsample_nearest2d.cpp (hunk 1 above)
Comment (llyfacebook) [reply aten/src/ATen/native/quantized/cpu/qupsample_nearest2d.cpp]: True, but there is no computation. Just memory access?

Comment (jamesr66a) [reply aten/src/ATen/native/quantized/cpu/qupsample_nearest2d.cpp]: Using vector instructions makes for cleaner and more efficient memory transfers. For example, see any stdlib's implementation of `memcpy`. It's hand-tuned vector assembly

Comment (jamesr66a) [reply aten/src/ATen/native/quantized/cpu/qupsample_nearest2d.cpp]: (you can also just call `memcpy` here)

File: aten/src/ATen/native/quantized/cpu/qupsample_nearest2d.cpp [hunk 4]
```diff
@@ -0,0 +1,167 @@
+#include <ATen/ATen.h>
+#include <ATen/core/op_registration/op_registration.h>
+#include <ATen/cpu/vec256/vec256.h>
+#include <ATen/native/SortingUtils.h>
+#include <ATen/native/TensorIterator.h>
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/qupsample_nearest2d.cpp]: You probably want to remove these four includes

**Review by jamesr66a — APPROVED** [review]
LGTM
